project/ShangHungWan/README.md

```diff
@@ -0,0 +1,121 @@
+
+![image 2: rectangles of R-tree](rtree-2.png)
+> image 2. Antonin Guttman. 1984. R-trees: a dynamic index structure for spatial searching. SIGMOD Rec. 14, 2 (June 1984), 47–57. <https://doi.org/10.1145/971697.602266>
+
+To handle round-off errors in the library. It will use epsilon. For example, it will allow a small tolerance threshold when it need to check overlap or compare the floating number.
+
+## Prospective Users
+
+Those who want to develop an application that needs to store/search geographical data.
+
+They can use both C++ and Python to include this library. C++ can directly include the library. On the other hand, this library also provides APIs for Python.
+
+## System Architecture
+
+### Interfaces
+
+This library will provide C++ interfaces:
+
+- RTRect RTRect::RTRect(double top, double right, double bottom, double left)
+- RTree RTree::RTree()
+- void RTree::insert(int uid, RTRect rect)
+- int RTree::nearest(RTRect rect)
+- int* RTree::intersection(RTRect rect)
+- void RTree::delete(int uid)
+
+And wrap it as Python APIs:
+
+- RTRect RTRect::RTRect(double top, double right, double bottom, double left)
+- RTree RTree::RTree()
+- void RTree::insert(int uid, RTRect rect)
+- int RTree::nearest(RTRect rect)
+- dict[int] RTree::intersection(RTRect rect)
+- void RTree::delete(int uid)
+
+### Testing
+
+This library will use randomly generated geographical dataset as the input to make sure its behavior is the same as expectation.
+
+Moreover, in order to make sure both C++ and Python's implementations are the same as our expectation. This library will contain unittest for both languages.
+
+## API Description
+
+Users can include rtreelib while compiling the code, for example:
+
+```bash
+g++ main.cpp -o main -lrtreelib
+```
+
+Furthermore, users who developing in Python can import it as well, for example:
+
+```python
+import rtreelib
+```
+
+About interfaces, please refer to [Interfaces](#interfaces).
+
+## Engineering Infrastructure
+
+### Build System
+
+- make
+- CI by GitHub Actions
+
+### Version control
+
+- Git
+- GitHub
+- GitHub Flow
+
+### Testing Framework
+
+- pytest (for Python)
+- Google Test (for C++)
+
+### Documentation
+
+- Markdown
+
+## Schedule
+
+|  Week   | Start Date | End Date | Description |
+|  ----  | ----  | ---- | ---- |
+| Planning phase  | 10/16 | 10/29 |  Keep editing and rectifying the proposal. |
+| Week 1  | 10/30 | 11/05 |  Read the paper and survey other references to understand R-tree well. |
+| Week 2  | 11/06 | 11/12 |  Implement `RTree::insert` and `RTRect`. |
+| Week 3  | 11/13 | 11/19 |  Implement `RTree::nearest` and `RTree::intersection`. |
+| Week 4  | 11/20 | 11/26 |  Implement `RTree::delete` and write unit tests that have about 50% code coverage. |
+| Week 5  | 11/27 | 12/03 |  Write unit tests that achieve 100% coverage. (ideally) |
```

Comment:
Better test as you write code.

Reply:
Changed the proposal; i plan to write the unit tests right after finish the component now👍